### DISCUSSION TITLE
android permissions example?

### DISCUSSION THREAD

**Discussion by Gibbz** [slint-ui/slint#5692] [Q&A]
I've got a basic app up and running in android with slint and rust. Is there any examples of setting up permissions in the same way android apps do? I'd like to add some Bluetooth support to my app....

- Setting up the permissions XML stuff
- Adding a popup box asking user for permission
- Applying the setting

**Comment by ogoffart** [comment]
Im afraid this kind of things is outside of the scope of Slint. You will have to find another rust library that does that, or reach into Java. 
I'm not aware of any example. Please let us know if you find something.
Maybe this would be the responsibility of xbuild?
Let us know as well if there is any enabler needed on the Slint platform required to make it work.

**Comment by Gibbz** [comment]
Okay thanks. I'm trying to find in the pipeline where this is done.
Are there any java rust bridge tools for android you know of?

This looks like it
https://github.com/jni-rs/jni-rs

**Reply by ogoffart** [reply]
> Are there any java rust bridge tools for android you know of?

Right, the JNI crate.

As an example, the i-slint-backend-android-activity also bundle some java. (It compiles it in its build.rs script)

**Reply by Gibbz** [reply]
What's the deference between the backends in slint?
backend-android-activity-05
backend-android-activity-06
I-slint-backend...

**Reply by ogoffart** [reply]
They are using a different version of the android-activity crate

**Comment by Gibbz** [comment]
I've been digging around for a few days trying to figure out how the android system works. And have it sused out.

Slint is using android-activity crate and have modified it to work for your purposes.
https://github.com/slint-ui/slint/tree/e0f3fd4168fa6ad5ed17a50910111ed72d1ae95e/internal/backends/android-activity
And custom java loader
https://github.com/slint-ui/slint/blob/e0f3fd4168fa6ad5ed17a50910111ed72d1ae95e/internal/backends/android-activity/java/SlintAndroidJavaHelper.java

Android activity examples here have the manifest file
https://github.com/rust-mobile/android-activity/tree/main/examples/agdk-mainloop/app/src/main

Slint doesn't have this manifest, but I need to use one to customise the android build, and permissions.
Also I need to modify the java file linked above with the permissions for android.

Is there a way I can essentially merge these two and put them in my project?

Slints android backend with extra manifest and code snippet.

**Comment by Gibbz** [comment]
Some samples here of adding Bluetooth permission to the manifest and java.
https://stackoverflow.com/questions/67722950/android-12-new-bluetooth-permissions

I'm wondering if it's possible to add a way to insert the permission code via rust, which is applied in the java? 

And a way to supply an android XML to slint?

**Comment by enomado** [comment]
@Gibbz any news ?

**Comment by Gibbz** [comment]
I gave up with it.
I was planning to grab the slint base android app and modify it. But figured my app was so simple I may as well just stick to the java.

**Comment by enomado** [comment]
@Gibbz sad to hear. Seems you are the one person who actually tried to use rust on android and get permissions.

I want draw fft spectrum from microphone and all i need is to request permission. I've done it with your snippet, call `requestPermissions` method on activity, but nothing happens.

Perhaps, I gave up too. Sorry for disturbing. Good luck.

**Comment by enomado** [comment]
it works 

```

fn req_perm(app: &winit::platform::android::activity::AndroidApp) -> anyhow::Result<()> {
    use jni::objects::JObject;
    use jni::sys::jobject;

    let ctx = ndk_context::android_context();
    let vm = unsafe { jni::JavaVM::from_raw(ctx.vm().cast()) }?;
    let mut env = vm.attach_current_thread()?;

    let record_string = env.new_string("android.permission.RECORD_AUDIO")?;

    let string_class = env.find_class("java/lang/String")?;
    let perms = env.new_object_array(1, string_class, record_string)?;

    let brr = (&perms).into();

    let activity = unsafe { JObject::from_raw(app.activity_as_ptr() as jobject) };

    let p = env
        .call_method(
            activity,
            "requestPermissions",
            "([Ljava/lang/String;I)V",
            &[brr, jni::objects::JValueGen::Int(1)],
        )
        .unwrap();

    Ok(())
}
```

**Reply by Gibbz** [reply]
Im having a quick play with this in my old app to see if i can get it going again. Ive a little trouble as im using slint so my main function is ```android_main(app: slint::android::AndroidApp)``` your app is ```req_perm(app: &winit::platform::android::activity::AndroidApp)``` What is the difference there? Are you using slint?

Also i dont see this function in your repo. Are you using it?

Edit2: is there a way to debug on android too?

**Comment by Gibbz** [comment]
Great, did you need to modify the XML?

**Reply by enomado** [reply]
just checked. manifest.yaml like this - works - https://github.com/rust-mobile/xbuild/issues/125#issuecomment-2424234343


```
android:
  manifest:
    uses_permission:
      - name: android.permission.RECORD_AUDIO

```

cargo.toml like this - NOT
```
[[package.metadata.android.uses_permission]]
name = "android.permission.RECORD_AUDIO"
```

you can check out my repo - https://github.com/enomado/waterfall

**Reply by Gibbz** [reply]
The manifest file is not working for me. See my post here if you can help:
https://github.com/rust-mobile/xbuild/issues/125#issuecomment-2558806400

looks like my file was just named wrong! after all this time! .yaml is correct, .yml is wrong

**Comment by Gibbz** [comment]
That's it! I never could find an example of the manifest yml for to work with either! 
Sounds like it's all coming together 👍
I'll close this then assuming it's all working 🙌

**Comment by wuwbobo2021** [comment]
Leaving a link here: <https://github.com/rust-mobile/android-activity/issues/174>, waiting for an answer from `android-activity` authors. One might do an experiment before it gets answered.

I found a Flutter library for reference: <https://github.com/baseflow/flutter-permission-handler>. However, it introduces `androidx` dependency, which should be avoided if a possible library for Rust should be compiled easily.

**Comment by rsyai-oos** [comment]
Hi, There, I'm working on egui on android, now I need to pick file using [egui_file_dialog](https://github.com/fluxxcode/egui-file-dialog.git), I initialized root path to `/storage/sdcard0`, and it can list folders in the directory, but can not list any files. 

I followd the discussions above, add manifest.yaml configuration and also tried @enomado req_perm(..) code. but it still can't work. 
@Gibbz Can you help to give any sugguestions?

![d44933b5ce87f1336d672d24d64d9bb8](https://github.com/user-attachments/assets/02db456a-6e38-4d39-aa99-a5704c24e16b)

**Reply by Gibbz** [reply]
I think I got permissions working. That's as far as I bothered to go. 

My source is here: 
https://github.com/bit-shift-io/bike-aid/blob/main/old/app-rs/src/android.rs

**Comment by wuwbobo2021** [comment]
Bold attempt of receiving result callback: <https://docs.rs/jni-min-helper/0.3.2/jni_min_helper/struct.PermissionRequest.html>, <https://github.com/rust-mobile/cargo-apk/pull/72>.

**Comment by wuwbobo2021** [comment]
New way of tracking permission request without requiring the use of `cargo-apk2`: <https://docs.rs/jni-min-helper/0.4.7/jni_min_helper/struct.PermissionRequest.html#method.request_with_activity>. `AndroidApp` is needed for getting the JNI global reference pointer of the native activity.

Note that this is unused in [tauri-plugin-blec](https://github.com/wuwbobo2021/tauri-plugin-blec) and [tauri-ble-example](https://github.com/wuwbobo2021/tauri-ble-example), diving into the Tauri mobile plugin system is a better approach there.

Usage:

```rust
        let activity_ptr = app.activity_as_ptr() as jni::sys::jobject;
        let activity = unsafe { env.as_cast_raw::<Global<JObject<'static>>>(&activity_ptr) }?;
        let request = PermissionRequest::request_with_activity(
            &activity,
            [
                "android.permission.ACCESS_FINE_LOCATION",
                "android.permission.READ_CALENDAR",
            ],
        )?;
```

This can be tried out with <https://crates.io/crates/android-ble>.

Also, `jni-min-helper` is looking for new maintainer(s).
